### PR TITLE
Mixed Precision Iterative Refinement example

### PR DESCRIPTION
**Pull request by upsj**
This PR adds an example of how to use the precision conversion functions to implement a Mixed Precision Iterative Refinement (MPIR) solver that uses a single-precision inner solver.

TODO:
- [x] Merge #521

### CODE REVIEW

File: examples/multiprecision-ir/doc/short-intro [hunk 1]
```diff
@@ -0,0 +1 @@
+The multiprecision iterative refinement solver example.
```
Comment (hartwiganzt) [inline examples/multiprecision-ir/doc/short-intro:1]: Mixed Precision Iterative Refinement (MPIR)

File: examples/multiprecision-ir/doc/tooltip [hunk 2]
```diff
@@ -0,0 +1 @@
+Manually implement a multiprecision iterative refinement method in Ginkgo. Solve a linear system.
```
Comment (hartwiganzt) [inline examples/multiprecision-ir/doc/tooltip:1]: mixed precision iterative refinement

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -12,6 +12,7 @@ add_subdirectory(ilu-preconditioned-solver)
 add_subdirectory(inverse-iteration)
 add_subdirectory(iterative-refinement)
 add_subdirectory(minimal-cuda-solver)
+add_subdirectory(multiprecision-ir)
```
Comment (hartwiganzt) [inline examples/CMakeLists.txt:15]: I know that this is a bigger change. But let's stick to the notation "mixed precision iterative refinement". That is well established - along with the abbreviation MPIR.

File: examples/multiprecision-ir/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+add_executable(multiprecision-ir multiprecision-ir.cpp)
+target_link_libraries(multiprecision-ir ginkgo)
+target_include_directories(multiprecision-ir PRIVATE ${PROJECT_SOURCE_DIR})
```
Comment (upsj) [inline examples/multiprecision-ir/CMakeLists.txt:3]:
```suggestion
add_executable(mixed-precision-ir mixed-precision-ir.cpp)
target_link_libraries(mixed-precision-ir ginkgo)
target_include_directories(mixed-precision-ir PRIVATE ${PROJECT_SOURCE_DIR})
```

File: examples/multiprecision-ir/build.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    LINK_FLAGS="-lginkgo -lginkgo_omp -lginkgo_cuda -lginkgo_reference -lginkgo_hip"
+else
+    LINK_FLAGS="-lginkgod -lginkgo_ompd -lginkgo_cudad -lginkgo_referenced -lginkgo_hipd"
+fi
+if [ -z "${CXX}" ]; then
+    CXX="c++"
+fi
+
+# build
+${CXX} -std=c++11 -o ${THIS_DIR}/multiprecision-ir \
```
Comment (upsj) [inline examples/multiprecision-ir/build.sh:35]:
```suggestion
${CXX} -std=c++11 -o ${THIS_DIR}/mixed-precision-ir \
```

File: examples/multiprecision-ir/doc/intro.dox [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<a name="Multiprecision Iterative Refinement"></a>
```
Comment (upsj) [inline examples/multiprecision-ir/doc/intro.dox:1]:
```suggestion
<a name="Mixed Precision Iterative Refinement"></a>
```

File: examples/multiprecision-ir/doc/intro.dox [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+<h1>This example manually implements a multiprecision iterative refinement solver.</h1>
```
Comment (upsj) [inline examples/multiprecision-ir/doc/intro.dox:2]:
```suggestion
<h1>This example manually implements a Mixed Precision Iterative Refinement solver.</h1>
```

File: examples/multiprecision-ir/doc/results.dox [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+Initial residual norm sqrt(r^T r):
+%%MatrixMarket matrix array real general
+1 1
+194.679
+Final residual norm sqrt(r^T r):
+%%MatrixMarket matrix array real general
+1 1
+4.23821e-11
+IR iteration count:     24
+IR execution time [ms]: 18.0692
```
Comment (yhmtsai) [inline examples/multiprecision-ir/doc/results.dox:15]: Is the result same as ir-example?

Comment (upsj) [reply examples/multiprecision-ir/doc/results.dox:15]: Yes, I should probably change that.

File: examples/CMakeLists.txt (hunk 3 above)
Comment (hartwiganzt) [reply examples/CMakeLists.txt:15]: Can you change the subdirectory name?

File: examples/mixed-precision-ir/doc/kind [hunk 9]
```diff
@@ -0,0 +1 @@
+basic
```
Comment (pratikvn) [inline examples/mixed-precision-ir/doc/kind:1]: I would maybe put this under a new heading mixed precision as we probably will have more examples doing mixed/adaptive precision. You can also change the adaptive-precision block jacobi to mixed-precision, I think

Comment (upsj) [reply examples/mixed-precision-ir/doc/kind:1]: That would mean adding a new style to `doc/scripts/examples.pl`, and adaptive-precision block-jacobi is already somewhat sensibly placed in the `preconditioner` group. How about we change this when we have more mixed precision examples?

Comment (pratikvn) [reply examples/mixed-precision-ir/doc/kind:1]: yeah, sure. I am okay with that.

[the remark this answers is not in the document]
Comment (tcojean) [reply examples/mixed-precision-ir/doc/kind:1]: Maybe in the meantime put this rather in "techniques" which is an alias for "advanced techniques" whereas basic is for "basic techniques". I would consider basic to be something which literally just calls Ginkgo interface. Although there is not much more, there is still more here.

**Pull request merged**